### PR TITLE
Housekeeping

### PR DESCRIPTION
**Pull request by kruschk**
There are a lot of changes here, but most of them are just making the imports more idiomatic. Please have a look at the commit messages for more details!

On a somewhat related note, it might not be a bad idea to run `cargo fmt -- --config max_width=80` (I think the default line widths are too long, so I like to set them to 80 characters). `cargo-fmt` usually does a pretty good job at formatting. I didn't want to go ahead and do this because the diffs would probably become an absolute mess!

### CODE REVIEW

**Comment by guygastineau** [comment]
@kruschk Thank you.  Yeah, I was being pretty lazy with formatting I think.  I agree 80 is better than 100 or 120 (I don't know what the default is).  I don't even remember if I have been using Clippy on my changes or not.

I will take a look tomorrow.  I have a lot of music obligations this evening.  Thanks again for your interest and participation!

**Comment by guygastineau** [comment]
I get the reasoning behind your second commit.  For large files with a whole lot going on I agree that the "idiomatic" approach makes sense.

There are a couple issues I find with being dogmatic about it though:

1. Rust in the wild has stabilized around conventions that are not faithfully represented in the book.  Natural and Computational languages both evolve with their community, and the Rust core teams attempt to remain "the central truth" about idiomatic usage of their language is impossible.

2. The end of the page acknowledges `glob`.  It does say to be careful, because the origin of a function can become unclear.  With the Rust LSP integrated into a text editor this is really not an issue in 2020 in my opinion.

3. I agree that is can still harm clarity.  `nom` is created to be a fast and light parser combinator DSL-like library.  It is not as much like a mini language as Parsec from Haskell, but I think that these mini-language functional parser libraries should be treated as mini-languages.  One benefit, then, is we can pull the names we need all the way in and use them like we are in a Rust-not-Rust ;)  Of course, some people in the Rust community would probably think this is an abhorrent mindset.

4. Given that none of the implementation files is very large in combination with point number 3, my preference is to leave most of the module prefixes out in the parsing modules.

5. I don't want you to think that I am completely stuck on that.  I think ultimately you made the imports look a lot nicer.  I am just distracted by the fact that some of the parsing functions are cluttered with so much contextual data that doesn't really add anything (feels like C++ or something).  I think the way forward is probably to use most the changes you have made.  Maybe the only times more should be brought in would even be inside functions (since luckily we can use `use` statements in the scope of a function).

-----

I will PR to your `kruschk:tidy` either when I get home from band practice tonight or tomorrow with the changes as I see them.  Thank you for holding off on the `fmt` changes. They are a good idea, but including them in another PR is the right way to go as you know.

I think your contribution here is a net positive, so thank you.  My resistance to literal adherence to Rust dogma does not undermine my appreciation for you work here.  I look forward to your feedback on the patch I will PR to your branch ;)

**Comment by kruschk** [comment]
Clippy only had only one recommendation, so you did pretty well if you're not using it! Anyway, it was my pleasure!

I think I see what you're getting at, and I can understand why you're not fully sold on the `module::function` approach. You hinted at the biggest issue I ran into in your second point: I was getting pretty lost trying to figure out the various functions' origins. As someone totally new to the codebase who also isn't familiar with `nom`, I like the dogmatic approach because of the extra context it gives. I do see how it can start to become verbose, though.

I know this PR was unwarranted, so I won't be offended if you don't accept all the changes. 😜 I just started fiddling to help me get acquainted with the codebase and got a bit carried away. I still need to study it more to understand, but I am slowly getting there.

Sounds good; enjoy your practice!

**Comment by guygastineau** [comment]
I probably make that second point in a too dismissive way most of the time anyway.  I probably spent too much time with Arch Linux, and I can't shake the RTFM mentality :rofl:   Also, it wasn't until I switched from Vim to Emacs that I learned about LSP integration with an editor in general.

In large part, I find the most important way to battle discoverability problems is to avoid glob using external crates (or internal crates if they are very big).  I mostly learned that from Haskell where depending on the circumstances you will just import a whole module (come directly into the namespace), do a qualified import with a prefix, or just import the exact definitions that you want from the given module.

Anyway, I have pulled your work, and I think I have found a compromise that will solve the issues that you made clear while still satisfying my sensibilities :smile: ~I should have a PR for you soon.~  I think I would prefer pushing straight to your branch (just less overhead that way).  Obviously we could rebase or whatever we need if that causes any issues with what we want in the end.  If you don't like that, then just let me know.  Otherwise, knowing that I plan to do that let's just make sure that neither of us does any force pushing to `kruschk/rexp:tidy`

File: src/parse/atom/mod.rs [hunk 1]
```diff
@@ -1,41 +1,38 @@
...
 }
 
-

-
 // String parsing
 
 fn string<'a>(i: &'a str) -> IResult<&'a str, String, VerboseError<&'a str>> {
+    use bytes::complete::tag;
+    use sequence::{terminated, preceded};
```
Comment (guygastineau) [inline src/parse/atom/mod.rs:35]: Here I am breaking with the styling I had before (as in I didn't include space next to the curly brackets).  I think this is what the Rust community expects.  Regardless, I want to format all of the this with `cargo fmt` after this PR is merged, so I am not worried about it right now.  I just wanted to be clear about the style inconsistency here, and that we are going to fix it automatically anyway ;)

**Comment by guygastineau** [comment]
@Kruschk please let me know if the compromise from my latest commit addresses your points adequately.

As an extra plus, getting rid of the namespace verbosity means that our lines appear to be less than 80 characters long again for the most part :smiling_imp:

**Comment by kruschk** [comment]
I know what you mean with the RTFM-attitude; I think it's still usually the way to go anyway. 😉 Yeah, that is certainly one of Vim's weak points. I'll have to give Emacs a try some day so I can have a more objective opinion on that holy war. I haven't even scratched the surface of Vim, though, and it still serves me quite well, so I'll probably stick with it for a while yet.

Thanks for your willingness to accommodate my concerns! Looking at these changes, I'm noticing there are some repetitive `use`s which take up a fair amount of vertical real estate that perhaps need not be occupied. I'm starting to think it may be best to just go back to the way you had it originally for function imports, and keep the other changes, Perhaps I've led us down a somewhat fruitless path with this PR, haha.

At the end of the day, the only thing I'd really like to avoid is glob imports. I think we're on the same page in that respect.

P.S. I'm still not the greatest with Git (or GitHub, for that matter), so if I'm missing something or doing something dumb, please let me know. 😅 For example, how is it that you were able to push directly to this branch? Is it because it's forked from a repository of which you're a maintainer, so you gain push permissions by default?

**Comment by kruschk** [comment]
Speaking of glob imports, perhaps we should clean this up while we're here? https://github.com/EKULibraries/rexp/blob/5502dab8e52c8d71692d0b578b91d26b7c46f8c7/src/parse/atom/mod.rs#L164

**Comment by guygastineau** [comment]
Glob imports are common for tests that are included in a submodule like that.  We don't need to make our lives too difficult.  Here is a link to the issue that made it so `clippy` would stop complaining about this pattern just this summer: [clippy being too pedantic](https://github.com/rust-lang/rust-clippy/issues/5554)

Concerning the vertical real-estate and the repetitive use statements in my patch on your PR:
* I don't actually mind the repetition or the extra lines especially as there is some variation.
* I think the goal should be legibility and discoverability, so lets think about this a bit more.  Since you are newer to the code base (and newer to Rust and `nom` in general), please let me know if you think the current approach would have been clearer for you when you first read through the code.  If it helps I see no harm in keeping it.

-----

Haha. I'll let you know if you break any "sacred rules of git" :P  Honeslty, your commit messages are nice and verbose.  I spend too much time doing solo dev, so you are already getting me to hold myself to a better standard of commit messages.  Good job picking up on the importance of commit messages!

I don't know if I can just straight up push to any branch on your fork.  I don't think I can.  I think GitHub specifically has an option that you (the PR author) can toggle to allow maintainers of the parent repo to edit your PR and its associated branch directly during the PR process.  I was originally thinking I would branch and track your `tidy` branch to make a PR, but I just felt like that was too much pomp and circumstance; thus I pushed straight to your branch.

Git is really fun and flexible software.  GitHub helps empower a lot of users at various levels (and it helps us share code more easily for collaboration), but often the real power and flexibility of `git` itself is obscured behind the GitHub specific mechanisms.  Understanding the program `git` makes using GitHub easier, but the reverse is not (at least as potently) so.

-----

Recap:
1. Get back to me after some reflection concerning legibility and discoverability.   If the parts of the `nom` API in use are clearer now, then we should go ahead and use this work.
2. Let me know if you think there are any other changes that really pertain to the PR.
3. Let me know if I answered your question about git and GitHub adequately.
4.  Know that I also used to be a vim user, so there is hope for you yet :rofl: Honeslty, vim is a great editor.  We often say that Emacs is great; if only it came with a decent text editor... :smiling_imp: I love it, but I also hate it everyday; I curse the existence of `elisp` daily, and I dream sweet dreams that emacs is running on guile with multithreading only to wake to this hellish reality of single-threaded Emacs :rofl:

**Comment by kruschk** [comment]
I suppose that's fair; good point.

It probably would have been clearer to a newbie, but I'm not entirely convinced the repetition is worth the convenience. My DRY-detector is going off! If you really prefer it this way, I won't complain; we can always change it in the future if we like.

---

That's good to hear, thanks! I read _Pro Git_ a little while back, which was very helpful, but I still forget things all the time and have to refer back to the book or Google how to do simple stuff. I'm slowly getting the hang of it, though.

Ohh, I see. That sounds like a nice feature to help reduce friction!

I certainly agree! GitHub is just a nice big wrapper around Git. 😝

---

1. Upon reflecting, I'm still not totally sold on it, but if we go for it and end up not liking it, we can always change it later.
2. Oh, I do have one other thing in mind. The current directory structure is below. Any thoughts on replacing `atom/mod.rs` with `atom.rs`?
```
rexp/src/
|-- expr.rs
|-- lib.rs
`-- parse
    |-- atom
    |   `-- mod.rs
    |-- mod.rs
    `-- quote.rs
```
3. You sure did, thanks!
4. Hahaha, I'm glad there is mutual respect here for our choice of editors. No editor can ever be perfect, unfortunately. 😢

**Comment by guygastineau** [comment]
Hmm, yeah I see what you mean.  I will take another pass where the most commonly used parser combinators (like `map`) are only `use`d once at the top.  I agree we wouldn;t want the import stuff to be too redundant, but DRY is not the really the same when it comes to the `used` statements.  I mean, if it adds clarity we shouldn't be concerned about a few repeated lines of imports.  That is not like having a bunch of repetitive ad hoc solutions to a problem that could be generalized throughout a codebase.

I think the reason why I have `atom` as a directory is because the module should be split into submodules.  Currently, the file has a big test embedded in it.  Either we will need to expose the inner working in order to test it in the external test (not considered the Rustiest way to do it), or we need to keep implementation test internal.  I assume we will stick with the latter.  I see a future where submodules (only files) for `atom` isolate the code from the re-export and unit test code.  Furthermore, I intend to change how Num is parsed.  In general I think it will be better for `atom` to be wrapper module with a unit test.  Character parsing has its own block that should be its own module.  So should strings, and nums probably.  I hope that clears up the reason for the filesystem anomaly.

**Comment by guygastineau** [comment]
FYI, I will do that second pass as soon as I have time (or you are welcome to do it if you want more commits to your name on this repo ;) ).  That way we can get this merged, and more fun work can get underway ;)

**Comment by kruschk** [comment]
Gotcha, I figured it was something along the lines of future-proofing, but thought I'd bring it up anyway.

Do you mean having another look at the `use`s? I think I'd better leave that one up to you. 😜

**Comment by guygastineau** [comment]
Hey, I have just been really swamped lately.  I will get the changes in, so this can be merged before the week is over.

**Comment by guygastineau** [comment]
Actually, I just looked over the embedded `use` statements again, and there is not very much repitition.  `combinator::map` is repeated maybe 3 times in `src/parse/atom/mod.rs`, but that is going to be multiple files in the future anyway...

So, I am ready to merge this!  I do think it has been a helpful PR, so thank you for your engagement with this project @kruschk !

I have realized a couple things going back over this code for the PR:

1. Like I mention in #3 I think number handling should be `dumb`er than it currently is.  I think it should probably match any of the legal numeric literals for a Scheme (although we could leave the onces that are preceeded by `#` for another pass if we want).  Basically, I don't think this library needs to handle producing some kind of int from Strings for consumers of the API.

2. I think that I jumped the gun with the quote, unquote, and splicing syntax parsing.  Those don't exist in "pure" symbolic expressions.  Rather they are reader extensions implemented in a Scheme (or Lisp) implementation.  I feel the same way about vectors and `#(`.  Thus, I think this library needs to remove the `Sexp::Vector`, and the `quote` types.  They should be replaced by some new type that represents a `sexp` list that has some junk stuck to the front of it.  That way `#()`, `'()`, `,()`, `,@()`, etc... would all be treated the same way.  Then it is up to subsequent passes to differentialte these forms, and even to locate the presence of quote and unquote(-splicing) prepended to variable names.  Let me know if you are interested in discussing this.

**Comment by guygastineau** [comment]
thank you for the PR!

**Pull request merged**